crates/wasmparser/src/readers/core/linking.rs

```diff
@@ -0,0 +1,405 @@
+            )),
+        }
+    }
+}
+
+/// Represents a data symbol.
+#[derive(Debug, Copy, Clone)]
+pub struct DefinedDataSymbol {
+    /// The index of the data segment.
+    pub index: u32,
+    /// The offset within the segment. Must be <= the segment's size.
+    pub offset: u32,
+    /// The size of the data, which can be zero. `offset + size` must be <= the segment's size.
+    pub size: u32,
+}
+
+impl<'a> FromReader<'a> for DefinedDataSymbol {
+    fn from_reader(reader: &mut BinaryReader<'a>) -> Result<Self> {
+        let index = reader.read_var_u32()?;
+        let offset = reader.read_var_u32()?;
+        let size = reader.read_var_u32()?;
+        Ok(Self {
+            index,
+            offset,
+            size,
+        })
+    }
+}
+
+/// Represents a subsection read from the linking custom section.
+#[derive(Debug, Clone)]
+pub enum Linking<'a> {
+    /// Extra metadata about the data segments.
+    SegmentInfo(SegmentMap<'a>),
+    /// A list of constructor functions to be called at startup.
+    InitFuncs(InitFuncMap<'a>),
+    /// The [COMDAT](https://llvm.org/docs/LangRef.html#comdats) groups of associated linking objects.
+    ComdatInfo(ComdatMap<'a>),
+    /// Extra information about the symbols present in the module.
+    SymbolTable(SymbolInfoMap<'a>),
+    /// An unknown [linking subsection](https://github.com/WebAssembly/tool-conventions/blob/main/Linking.md#linking-metadata-section).
+    Unknown {
+        /// The identifier for this subsection.
+        ty: u8,
+        /// The contents of this subsection.
+        data: &'a [u8],
+        /// The range of bytes, relative to the start of the original data
+        /// stream, that the contents of this subsection reside in.
+        range: Range<usize>,
+    },
+}
+
+impl<'a> Subsection<'a> for Linking<'a> {
+    fn from_reader(id: u8, reader: BinaryReader<'a>) -> Result<Self> {
+        let data = reader.remaining_buffer();
+        let offset = reader.original_position();
+        Ok(match id {
+            5 => Self::SegmentInfo(SegmentMap::new(data, offset)?),
+            6 => Self::InitFuncs(InitFuncMap::new(data, offset)?),
+            7 => Self::ComdatInfo(ComdatMap::new(data, offset)?),
+            8 => Self::SymbolTable(SymbolInfoMap::new(data, offset)?),
+            ty => Self::Unknown {
+                ty,
+                data,
+                range: offset..offset + data.len(),
+            },
+        })
+    }
+}
+
+impl<'a> LinkingSectionReader<'a> {
+    /// Creates a new reader for the linking section contents starting at
+    /// `offset` within the original wasm file.
+    pub fn new(data: &'a [u8], offset: usize) -> Result<Self> {
+        let mut reader = BinaryReader::new_with_offset(data, offset);
+        let range = reader.range();
+
+        let version = reader.read_var_u32()?;
+        if version != 2 {
+            return Err(BinaryReaderError::new(
+                format!("unsupported linking section version: {}", version),
```

Comment:
Do you think it'd be useful to provide an easier way for users to differentiate between
* the `linking` section is there, but its contents are malformed — we report an error and abort
* the `linking` section is there, but we don't support parsing this version — we print a warning and continue

Reply:
Up to you! Currently `wasmparser` doesn't do a great job of making errors introspectable though, right now they're all simply treated as "fatal please print somewhere and move on" as opposed to being specifically actionable one way or another.